cranelift-codegen/src/regalloc/branch_splitting.rs

```diff
@@ -0,0 +1,223 @@
+use crate::dominator_tree::DominatorTree;
+use crate::flowgraph::ControlFlowGraph;
+use crate::ir::{Ebb, Function, Inst, InstBuilder, InstructionData, Opcode, Value};
+use crate::isa::TargetIsa;
+use crate::topo_order::TopoOrder;
+
+pub fn run(
+    isa: &TargetIsa,
+    func: &mut Function,
+    cfg: &mut ControlFlowGraph,
+    domtree: &mut DominatorTree,
+    topo: &mut TopoOrder,
+) {
+    let mut ctx = Context {
+        new_blocks: false,
+        cur: EncCursor::new(func, isa),
+        domtree,
+        topo,
+        cfg,
+    };
+    ctx.run()
+}
+
+struct Context<'a> {
+    // True if new blocks were inserted.
+    new_blocks: bool,
+
+    // Current instruction as well as reference to function and ISA.
+    cur: EncCursor<'a>,
+
+    // References to contextual data structures we need.
+    domtree: &'a mut DominatorTree,
+    topo: &'a mut TopoOrder,
+    cfg: &'a mut ControlFlowGraph,
+}
+
+impl<'a> Context<'a> {
+    fn run(&mut self) {
+        // Any ebb order will do.
+        self.topo.reset(self.cur.func.layout.ebbs());
+        while let Some(ebb) = self.topo.next(&self.cur.func.layout, self.domtree) {
+            // Branches can only be at the last or second to last position in an extended basic
+            // block.
+            self.cur.goto_last_inst(ebb);
+            let terminator_inst = self
+                .cur
+                .current_inst()
+                .expect("missing terminator instruction");
+            match self.cur.prev_inst() {
+                Some(inst) => {
+                    let opcode = self.cur.func.dfg[inst].opcode();
+                    if opcode.is_branch() {
+                        self.visit_conditional_branch(inst, opcode);
+                    } else {
+                        continue;
+                    }
+                }
+                // If there is only a terminator instruction, then there is no need to create a new
+                // block.
+                None => continue,
+            };
+            self.cur.goto_inst(terminator_inst);
+            self.visit_terminator_branch(terminator_inst);
+        }
+
+        // If blocks were added the cfg and domtree are inconsistent and must be recomputed.
+        if self.new_blocks {
+            self.cfg.compute(&self.cur.func);
+            self.domtree.compute(&self.cur.func, self.cfg);
+        }
+    }
+
+    fn visit_conditional_branch(&mut self, inst: Inst, opcode: Opcode) {
+        let target = match self.cur.func.dfg[inst] {
+            InstructionData::Branch { destination, .. }
+            | InstructionData::BranchIcmp { destination, .. }
+            | InstructionData::BranchInt { destination, .. }
+            | InstructionData::BranchFloat { destination, .. } => destination,
+            _ => panic!("Unexpected instruction in classify_branch"),
+        };
+
+        // If there are any parameters, split the edge.
+        if self.should_split_edge(inst, target) {
```

Comment:
Pretty silly nit: instead of indenting the whole (long!) function, can just be:

```rust
if !self.should_split_edge(inst, target) {
    return;
}
```